p2p/src/peer_manager/peerdb/address_tables/mod.rs

```diff
@@ -0,0 +1,190 @@
+//!
+//! Each table is represented by the `Table` struct, which consists of a fixed number of buckets
+//! each of which contains a fixed number of slots; the maximum number of addresses that a table
+//! can hold is the product of those numbers.
+//! The indices of a bucket and of a slot that an address will be put into are determined by
+//! hashing the address (to be more precise, different parts of the address are hashed separately
+//! and then combined together; this allows to ensure, for example, that addresses from a certain
+//! "address group" can be spread over only a very limited number of buckets).
+//!
+//! `Table` has a constant memory overhead equal to `buckets_count*bucket_size*size_of(u32)`,
+//! so using big numbers for the bucket size and count is not a good idea. On the other hand,
+//! choosing small values might not be a good idea either, because potentially useful addresses
+//! may start evicting each other due to hash collisions.
+
+use std::hash::{Hash, Hasher};
+
+use crypto::random::{make_pseudo_rng, Rng};
+use p2p_types::socket_address::SocketAddress;
+use serialization::{Decode, Encode};
+
+use self::table::Table;
+
+use super::config::PeerDbConfig;
+
+#[macro_use]
+pub mod table;
+
+/// The random key that is used to randomize bucket selection inside the tables.
+///
+/// It will be randomly generated once and stored in the peer db.
+// TODO: is "salt" a better name for this?
+#[derive(Hash, Encode, Decode, Copy, Clone)]
+pub struct RandomKey(u64);
+
+impl RandomKey {
+    pub fn new_random() -> Self {
+        Self::new_random_with_rng(&mut make_pseudo_rng())
+    }
+
+    pub fn new_random_with_rng<R: Rng>(rng: &mut R) -> Self {
+        RandomKey(rng.gen::<u64>())
+    }
+
+    pub fn mix_with<T: Hash>(&self, data: T) -> RandomKey {
+        RandomKey(calc_hash!(self.0, data))
+    }
+}
+
+pub struct AddressTables {
+    new_addr_table: Table,
+    tried_addr_table: Table,
+}
+
+impl AddressTables {
+    pub fn new(random_key: RandomKey, peerdb_config: &PeerDbConfig) -> Self {
+        let new_addr_table = Table::new(
+            *peerdb_config.new_addr_table_bucket_count,
+            *peerdb_config.addr_tables_bucket_size,
+            random_key.mix_with('n'),
+        );
+        let tried_addr_table = Table::new(
+            *peerdb_config.tried_addr_table_bucket_count,
+            *peerdb_config.addr_tables_bucket_size,
+            random_key.mix_with('t'),
+        );
+
+        Self {
+            new_addr_table,
+            tried_addr_table,
+        }
+    }
+
+    /// Remove the specified address from "new" and put it into "tried".
+    /// If an older "tried" address exists at the destination, move it into "new".
+    /// The return value is `Option<(old_tried, Option<old_new>)>`, where old_tried is the address
+    /// that was moved from 'tried' to 'new' and old_new is the address that was removed from
+    /// 'new' due to its place being taken by old_tried.
```

Comment:
I decided to make it a struct instead:
```
/// Optional side effects of `move_to_tried` (the additional addresses that it had to move).
pub struct MoveToTriedSideEffects {
    /// A "tried" address that had to be moved to "new".
    pub addr_moved_to_new: SocketAddress,
    /// The address that was removed from "new" and is no longer in the tables.
    pub discarded_new_addr: Option<SocketAddress>,
}
```
And `move_to_tried` returns `Option<MoveToTriedSideEffects>`.

I also changed the result of `move_to_new`, which now returns `MoveToNewOutcome`:
```
pub enum MoveToNewOutcome {
    /// The address was successfully put into the "new" table.
    Succeeded {
        /// The previous address that existed at the corresponding slot.
        prev_addr: Option<SocketAddress>,
    },
    /// `move_to_new` didn't do anything, because the address that already exists
    /// at the corresponding slot is better. 
    Cancelled,
}
```